File: src/ns/mixer-device.ts
import { Ableton } from "..";
import { Namespace } from ".";
import { DeviceParameter, RawDeviceParameter } from "./device-parameter";

export enum PanningMode {
  Stereo,
  StereoSplit,
}

export enum CrossfadeAssignment {
  A,
  None,
  B,
}

export interface GettableProperties {
  crossfade_assign: CrossfadeAssignment;
  crossfader: RawDeviceParameter;
  cue_volume: RawDeviceParameter;
  left_split_stereo: RawDeviceParameter;
  panning: RawDeviceParameter;
  panning_mode: PanningMode;
  right_split_stereo: RawDeviceParameter;
  sends: RawDeviceParameter[];
  song_tempo: RawDeviceParameter;
  track_activator: RawDeviceParameter;
  volume: RawDeviceParameter;
}

export interface TransformedProperties {
  crossfader: DeviceParameter;
  cue_volume: DeviceParameter;
  left_split_stereo: DeviceParameter;
  panning: DeviceParameter;
  right_split_stereo: DeviceParameter;
  sends: DeviceParameter[];
  song_tempo: DeviceParameter;
  track_activator: DeviceParameter;
  volume: DeviceParameter;
}

export interface SettableProperties {
  crossfade_assign: CrossfadeAssignment;
  panning_mode: string;
}

export interface ObservableProperties {
  crossfade_assign: CrossfadeAssignment;
  panning_mode: string;
  sends: RawDeviceParameter[];
}

export interface RawMixerDevice {
  id: string;
  volume: string;
}

export class MixerDevice extends Namespace<
  GettableProperties,
  TransformedProperties,
  SettableProperties,
  ObservableProperties
> {
  constructor(
    ableton: Ableton,
    public raw: RawMixerDevice,
  ) {
    super(ableton, "mixer-device", raw.id);

    this.transformers = {
      crossfader: (v) => new DeviceParameter(ableton, v),
      cue_volume: (v) => new DeviceParameter(ableton, v),
      left_split_stereo: (v) => new DeviceParameter(ableton, v),
      panning: (v) => new DeviceParameter(ableton, v),
      right_split_stereo: (v) => new DeviceParameter(ableton, v),
      sends: (v) => v.map((s) => new DeviceParameter(ableton, s)),
      song_tempo: (v) => new DeviceParameter(ableton, v),
      track_activator: (v) => new DeviceParameter(ableton, v),
      volume: (v) => new DeviceParameter(ableton, v),
    };
  }
}

File: midi-script/Namespaces/MixerDevice.py
from __future__ import absolute_import

import Live

from .DeviceParameter import DeviceParameter
from .Interface import Interface


class MixerDevice(Interface):
    @staticmethod
    def serialize_mixer_device(mixer_device):
        if mixer_device is None:
            return None

        device_id = Interface.save_obj(mixer_device)
        return {"id": device_id, "volume": mixer_device.volume}

    def __init__(self, c_instance, socket):
        super().__init__(c_instance, socket)

    def get_crossfade_assign(self, ns):
        return str(ns.crossfade_assign)

    def set_crossfade_assign(self, ns, value):
        ns.crossfade_assign = getattr(
            Live.MixerDevice.MixerDevice.crossfade_assignments, str(value)
        )

    def get_panning_mode(self, ns):
        return str(ns.panning_mode)

    def set_panning_mode(self, ns, value):
        ns.panning_mode = getattr(
            Live.MixerDevice.MixerDevice.panning_modes, str(value)
        )

    def get_crossfader(self, ns):
        return DeviceParameter.serialize_device_parameter(ns.crossfader)

    def get_cue_volume(self, ns):
        return DeviceParameter.serialize_device_parameter(ns.cue_volume)

    def get_left_split_stereo(self, ns):
        return DeviceParameter.serialize_device_parameter(ns.left_split_stereo)

    def get_panning(self, ns):
        return DeviceParameter.serialize_device_parameter(ns.panning)

    def get_right_split_stereo(self, ns):
        return DeviceParameter.serialize_device_parameter(ns.right_split_stereo)

    def get_sends(self, ns):
        return map(DeviceParameter.serialize_device_parameter, ns.sends)

    def get_song_tempo(self, ns):
        return DeviceParameter.serialize_device_parameter(ns.song_tempo)

    def get_track_activator(self, ns):
        return DeviceParameter.serialize_device_parameter(ns.track_activator)

    def get_volume(self, ns):
        return DeviceParameter.serialize_device_parameter(ns.volume)
